src/modalities/models/coca/multi_modal_decoder.py

```diff
@@ -0,0 +1,116 @@
+        n_embd: int,
+        bias: bool,
+        epsilon: float,
+        activation: ActivationType,
+        n_head: int,
+        dropout: float,
+        ffn_hidden: int,
+        attention_config: AttentionConfig = None,
+        attention_type: AttentionType = AttentionType.NON_CAUSAL_ATTENTION,
+        with_context: bool = True,
+    ):
+        super().__init__()
+        self.with_context = with_context
+
+        if activation == ActivationType.GELU:
+            mlp = partial(MLP, in_features=n_embd, hidden_features=ffn_hidden, bias=bias, dropout=dropout)
+        elif activation == ActivationType.FUSED_SWIGLU:
+            mlp = partial(xops.SwiGLU, in_features=n_embd, hidden_features=ffn_hidden, bias=bias)
+        else:
+            raise NotImplementedError(f"activation type {activation} not implemented")
+
+        self.ln_1 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+        self.attn = MultiHeadAttention(
+            n_embd=n_embd, n_head=n_head, bias=bias, attention_config=attention_config, attention_type=attention_type
+        )
+        self.ln_2 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+        self.mlp = mlp()
+
+        if self.with_context:
+            self.ln_3 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+            self.cross_attn = MultiHeadAttention(
+                n_embd=n_embd,
+                n_head=n_head,
+                bias=bias,
+                attention_config=attention_config,
+                attention_type=AttentionType.CROSS_ATTENTION,
+            )
+            self.ln_4 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+            self.mlp_2 = mlp()
+
+    def forward(self, x: torch.Tensor, context: torch.Tensor = None) -> torch.Tensor:
+        x = x + self.attn(self.ln_1(x))
+        x = x + self.mlp(self.ln_2(x))
+        if self.with_context:
+            x = x + self.cross_attn(self.ln_3(x), context=context)
+            x = x + self.mlp_2(self.ln_4(x))
+        return x
+
+
+class MultiModalDecoder(NNModel):
+    def __init__(
+        self,
+        sample_key: str,
+        prediction_key: str,
+        block_size: int,
+        vocab_size: int,
+        n_layer: int,
+        n_head: int,
+        n_embd: int,
+        ffn_hidden: int,
+        dropout: float,
+        bias: bool,
+        attention_config: AttentionConfig,
+        activation: ActivationType,
+        epsilon: float,
+    ):
+        super().__init__()
+        self.sample_key = sample_key
+        self.prediction_key = prediction_key
+        self.block_size = block_size
+
+        self.transformer = nn.ModuleDict(
+            dict(
+                h=nn.ModuleList(
+                    [
+                        MultiModalBlock(
+                            n_embd=n_embd,
+                            bias=bias,
+                            epsilon=epsilon,
+                            activation=activation,
+                            attention_config=attention_config,

```

Comment:
`attention_config` should maybe be moved to last position, such that the order of arguments matches the one in `MultiModalBlock`'s `__init__`

Reply:
https://github.com/Modalities/modalities/pull/16/commits/be9e01d2156f7fc00f681286c9f202bb77afd5ec